/*
*  Copyright (c) 2020, The OpenThread Authors.
*  Copyright 2025 NXP
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions are met:
*  1. Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*  2. Redistributions in binary form must reproduce the above copyright
*     notice, this list of conditions and the following disclaimer in the
*     documentation and/or other materials provided with the distribution.
*  3. Neither the name of the copyright holder nor the
*     names of its contributors may be used to endorse or promote products
*     derived from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
*  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
*  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
*  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
*  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
*  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
*  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
*  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
*  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*/

/*
* Header for patch information.
* Description           : It is required to restart dhcpcd on PD state change.
* Version               : v001
* Mandatory/Optional    : Mandatory
* Impact if not applied : /etc/default/otbr-agent file can't beneficiate of DTB gpio naming.
*/

Upstream-Status: Inappropriate  [platform specific changes]

diff --git a/script/reference-device/dhcp6_pd_daemon.py b/script/reference-device/dhcp6_pd_daemon.py
index fa386ed1..5be4f91f 100755
--- a/script/reference-device/dhcp6_pd_daemon.py
+++ b/script/reference-device/dhcp6_pd_daemon.py
@@ -56,7 +56,7 @@ def restart_dhcpcd_service(config_path):
     try:
         subprocess.run(["cp", config_path, DHCP_CONFIG_PATH], check=True)
         subprocess.run(["systemctl", "daemon-reload"], check=True)
-        subprocess.run(["service", "dhcpcd", "restart"], check=True)
+        subprocess.run(["systemctl", "restart", "dhcpcd"], check=True)
         logging.info(
             f"Successfully restarted dhcpcd service with {config_path}.")
     except subprocess.CalledProcessError as e:
@@ -122,8 +122,10 @@ def main():
     # By restarting dhcpcd here, we ensure it runs after network.target is active, allowing
     # radvd to start correctly and dhcpcd to configure the interface.
     try:
-        subprocess.run(["systemctl", "reload-or-restart", "dhcpcd"], check=True)
-        logging.info("Successfully restarting dhcpcd service.")
+        logging.info(f"intended_dhcp6pd_state: {intended_dhcp6pd_state}")
+        if intended_dhcp6pd_state is not None:
+            subprocess.run(["systemctl", "reload-or-restart", "dhcpcd"], check=True)
+            logging.info("Successfully restarting dhcpcd service.")
     except subprocess.CalledProcessError as e:
         logging.error(f"Error restarting dhcpcd service: {e}")
 
diff --git a/src/agent/CMakeLists.txt b/src/agent/CMakeLists.txt
index 4c27f412..b74f8d44 100644
--- a/src/agent/CMakeLists.txt
+++ b/src/agent/CMakeLists.txt
@@ -109,6 +109,16 @@ install(FILES ${CMAKE_CURRENT_BINARY_DIR}/imx_otbr_gpiochip_spidev_cfg.sh
     DESTINATION ${OTBR_SYSTEMD_UNIT_DIR}
 )
 
+install(FILES ${CMAKE_SOURCE_DIR}/script/reference-device/dhcp6_pd_daemon.py
+    DESTINATION ${OTBR_SYSTEMD_UNIT_DIR}
+)
+install(FILES ${CMAKE_SOURCE_DIR}/script/reference-device/dhcpcd.enter-hook
+    DESTINATION ${OTBR_SYSTEMD_UNIT_DIR}
+)
+install(FILES ${CMAKE_SOURCE_DIR}/script/reference-device/dhcpcd.exit-hook
+    DESTINATION ${OTBR_SYSTEMD_UNIT_DIR}
+)
+
 pkg_check_modules(LIBSYSTEMD libsystemd)
 if(LIBSYSTEMD_FOUND)
     target_compile_definitions(otbr-config INTERFACE "HAVE_LIBSYSTEMD=1")

